File: infrastructure/db-core/src/main/java/com/nmh/commerce/coupon/CouponStockRepositoryImpl.java
package com.nmh.commerce.coupon;

import jakarta.persistence.OptimisticLockException;
import org.hibernate.StaleObjectStateException;
import org.springframework.stereotype.Repository;

import java.util.Optional;

@Repository
public class CouponStockRepositoryImpl implements CouponStockRepository {

    private final CouponStockJpaRepository couponStockJpaRepository;

    public CouponStockRepositoryImpl(CouponStockJpaRepository couponStockJpaRepository) {
        this.couponStockJpaRepository = couponStockJpaRepository;
    }

    @Override
    public CouponStock save(CouponStock couponStock) {
        return couponStockJpaRepository.save(CouponStockEntity.from(couponStock)).toDomain();
    }

    @Override
    public Optional<CouponStock> findByCouponId(Long couponId) {
        return couponStockJpaRepository.findByCouponId(couponId)
            .map(CouponStockEntity::toDomain);
    }

    @Override
    public CouponStock update(CouponStock couponStock) {
        try {
            return couponStockJpaRepository.save(CouponStockEntity.from(couponStock)).toDomain();
        } catch (OptimisticLockException | StaleObjectStateException e) {
            throw new IllegalStateException("Failed to update coupon stock", e);
        } catch (Exception e) {
            throw new RuntimeException("An unexpected error occurred", e);
        }
    }

}

File: infrastructure/db-core/src/main/kotlin/com/nmh/commerce/coupon/CouponStockRepositoryImpl.kt
package com.nmh.commerce.coupon

import jakarta.persistence.OptimisticLockException
import org.hibernate.StaleObjectStateException
import org.springframework.stereotype.Repository

@Repository
class CouponStockRepositoryImpl(private val couponStockJpaRepository: CouponStockJpaRepository) :
    CouponStockRepository {
    override fun save(couponStock: CouponStock): CouponStock {
        return couponStockJpaRepository.save(CouponStockEntity.from(couponStock)).toDomain()
    }

    override fun findByCouponId(couponId: Long): CouponStock {
        return couponStockJpaRepository.findByCouponId(couponId)
            ?.toDomain()
            ?: throw NoSuchElementException("Coupon stock not found for couponId: $couponId")
    }

    override fun update(couponStock: CouponStock): CouponStock {
        try {
            return couponStockJpaRepository.save<CouponStockEntity>(CouponStockEntity.from(couponStock)).toDomain()
        } catch (e: OptimisticLockException) {
            throw IllegalStateException("Failed to update coupon stock", e)
        } catch (e: StaleObjectStateException) {
            throw IllegalStateException("Failed to update coupon stock", e)
        } catch (e: Exception) {
            throw RuntimeException("An unexpected error occurred", e)
        }
    }
}
